File: 1396-design-underground-system/1396-design-underground-system.java
class UndergroundSystem {
    HashMap<Integer, Trip> currentTrips;
    HashMap<String, List<Trip>> completedTrips;
    
    
    public UndergroundSystem() {
        this.currentTrips = new HashMap<>();
        this.completedTrips = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        if(this.currentTrips.containsKey(id))
            return;
        
        this.currentTrips.put(id, new Trip(id, stationName, t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        if(!this.currentTrips.containsKey(id))
            return;
        
        Trip curr = currentTrips.get(id);
        currentTrips.remove(id);
        curr.end = stationName;
        curr.endTime = t;
        
        completedTrips.putIfAbsent(curr.start, new ArrayList<>());
        completedTrips.get(curr.start).add(curr);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        List<Trip> trips = completedTrips.get(startStation);
        if(trips == null || trips.size() == 0)
            return 0;
        
        int count = 0;
        int total = 0;
        
        for(Trip trip : trips){
            if(trip.end.equals(endStation)){
                total += (trip.endTime - trip.startTime);
                ++count;
            }
        }
        
        return total / (double) count;
    }
}

class Trip {
    int id;
    String start;
    int startTime;
    String end;
    int endTime;
    
    public Trip(int id, String stationName, int t){
        this.id = id;
        this.start = stationName;
        this.startTime = t;
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: 1396-design-underground-system/1396-design-underground-system.py
class UndergroundSystem:

    def __init__(self):
        self.cust_map = {}
        self.stat_map = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.cust_map[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        start, start_t = self.cust_map[id]
        del self.cust_map[id]
        if (start, stationName) not in self.stat_map:
            self.stat_map[(start, stationName)] = []
        self.stat_map[(start, stationName)].append(t - start_t)
        

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        total = 0
        count = 0
        for t in self.stat_map[(startStation, endStation)]:
            total += t
            count += 1
        
        return total / count if count else 0


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)